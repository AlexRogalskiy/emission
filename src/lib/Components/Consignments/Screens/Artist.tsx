import React from "react"

import ConsignmentBG from "../Components/ConsignmentBG"
import DoneButton from "../Components/DoneButton"
import { SearchResults } from "../Components/SearchResults"

import { ArtistResult, ConsignmentSetup } from "../index"

import { Dimensions, Route, View, ViewProperties } from "react-native"
import NavigatorIOS from "react-native-navigator-ios"
import metaphysics from "../../../metaphysics"

interface ArtistSearchResponse {
  match_artist: ArtistResult[]
}

interface Props extends ConsignmentSetup, ViewProperties {
  navigator: NavigatorIOS
  route: Route
  updateWithArtist?: (result: ArtistResult) => void
}

interface State {
  query: string
  searching: boolean
  results: ArtistResult[] | null
}

export default class Artist extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props)
    this.state = {
      query: null as any,
      searching: false,
      results: null as any,
    }
  }

  doneTapped = () => {
    this.props.navigator.pop()
  }

  artistSelected = (result: ArtistResult) => {
    this.props.updateWithArtist(result)
    this.props.navigator.pop()
  }

  textChanged = async (text: string) => {
    this.setState({ query: text, searching: text.length > 0 })
    this.searchForQuery(text)
  }

  searchForQuery = async (query: string) => {
    const results = await metaphysics<ArtistSearchResponse>(`
      {
        match_artist(term: "${query}") {
          internalID
          name
          image {
            url
          }
        }
      }
    `)
    this.setState({ results: results.match_artist, searching: false })
  }

  render() {
    const isPad = Dimensions.get("window").width > 700

    return (
      <ConsignmentBG>
        <DoneButton onPress={this.doneTapped}>
          <View
            style={{
              alignContent: "center",
              justifyContent: isPad ? "center" : "flex-end",
              flexGrow: 1,
              marginLeft: 20,
              marginRight: 20,
            }}
          >
            <SearchResults
              results={this.state.results}
              query={this.state.query}
              placeholder="Artist/Designer Name"
              noResultsMessage="Unfortunately we are not accepting consignments for works by"
              onChangeText={this.textChanged}
              searching={this.state.searching}
              resultSelected={this.artistSelected}
            />
          </View>
        </DoneButton>
      </ConsignmentBG>
    )
  }
}
