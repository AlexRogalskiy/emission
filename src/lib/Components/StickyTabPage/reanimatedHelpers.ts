import { useMemo, useRef, useState } from "react"
import Animated from "react-native-reanimated"

/**
 * returns a stable Animated.Value instance which starts off with the
 * given number. Note that the initialization parameter will be ignored
 * on subsequent renders
 * @param init
 */
export function useAnimatedValue(init: number) {
  return useMemo(() => {
    return new Animated.Value(init)
  }, [])
}

/**
 * returns a function which can asynchronously read the value of some native animated nodes
 * This is useful for values that change frequently on the native side but which you only
 * want to read occasionally on the JS side. It helps you avoid the perf hit of sending change
 * events over the bridge.
 * @param vals the animated vals to make the reader function for
 * @example
 * const scrollOffset = useMemo(() => new Animated.Value(0), [])
 * const readVals = useValueReader({scrollOffset})
 * // later, e.g. in a callback
 * const {scrollOffset} = await readVals()
 * console.log(scrollOffset) // => 632
 */
export function useValueReader<T extends { [k: string]: Animated.Adaptable<number> }>(
  props: T
): () => Promise<{ [k in keyof T]: number }> {
  // this works by running some reanimated code every time an 'epoch' value
  // is incremented. That code calls a callback with the current values
  // to resolve a promise set up for the consumer
  const epochRef = useRef(0)
  const epoch = useAnimatedValue(0)
  const lastEpoch = useAnimatedValue(0)

  const readCallback = useRef<(vals: ReadonlyArray<number>) => void>()

  const keys = useMemo(() => {
    return Object.keys(props)
  }, [props])

  const vals = useMemo(() => {
    return keys.map(k => props[k])
  }, [keys])

  Animated.useCode(
    () =>
      Animated.cond(Animated.neq(epoch, lastEpoch), [
        Animated.set(lastEpoch, epoch),
        Animated.call([...vals], vs => {
          const cb = readCallback.current
          readCallback.current = null
          result.current = null
          cb(
            keys.reduce((acc, k, i) => {
              acc[k] = vs[i]
              return acc
            }, {} as any)
          )
        }),
      ]),
    [vals, keys]
  )

  const result = useRef<Promise<any>>()

  return () => {
    if (!result.current) {
      result.current = new Promise(resolve => {
        readCallback.current = resolve
        epochRef.current += 1
        epoch.setValue(epochRef.current)
      })
    }
    return result.current
  }
}

export function useNativeValue(node: Animated.Node<number>, init: number): number {
  const [state, setState] = useState(init)
  Animated.useCode(() => Animated.call([node], ([val]) => setState(val)), [])
  return state
}

// Takes an animated driver value and outputs a value which follows the
// driver but smoothed using a spring damping technique
export function springDampen(val: Animated.Node<number>, customConfig: Partial<Animated.SpringConfig> = {}) {
  const {
    Clock,
    Value,
    neq,
    sub,
    add,
    and,
    block,
    cond,
    set,
    spring,
    not,
    clockRunning,
    startClock,
    stopClock,
  } = Animated
  const clock = new Clock()
  const dampener = new Value(0)
  const output = add(val, dampener)
  const lastVal = new Value(0)
  const firstTime = new Value(1)
  const state = {
    finished: new Value(0),
    position: dampener,
    velocity: new Value(0),
    time: new Value(0),
  }

  const config: Animated.SpringConfig = {
    stiffness: new Value(180),
    mass: new Value(1),
    damping: new Value(18),
    overshootClamping: false,
    restSpeedThreshold: 0.001,
    restDisplacementThreshold: 0.001,
    ...customConfig,
    toValue: 0,
  }

  return block([
    cond(firstTime, [set(firstTime, 0), set(lastVal, output)]),
    set(dampener, add(dampener, sub(lastVal, val))),
    set(lastVal, val),
    cond(and(neq(0, dampener), not(clockRunning(clock))), [
      startClock(clock),
      set(state.finished, 0),
      set(state.velocity, 0),
      set(state.time, 0),
    ]),
    cond(clockRunning(clock), [
      spring(clock, state, config),
      // if the animation is over we stop the clock
      cond(state.finished, stopClock(clock)),
    ]),
    output,
  ])
}
