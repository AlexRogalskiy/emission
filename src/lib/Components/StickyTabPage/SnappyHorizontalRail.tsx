import { useScreenDimensions } from "lib/utils/useScreenDimensions"
import React, { useImperativeHandle, useRef } from "react"
import Animated from "react-native-reanimated"
import { useAnimatedValue } from "./reanimatedHelpers"

export interface SnappyHorizontalRail {
  setOffset(offset: number): void
}
export const SnappyHorizontalRail = React.forwardRef<
  SnappyHorizontalRail,
  React.PropsWithChildren<{ initialOffset?: number }>
>(({ children, initialOffset = 0 }, ref) => {
  const { width } = useScreenDimensions()

  const currentOffset = useAnimatedValue(-initialOffset)
  const currentAnimation = useRef<Animated.BackwardCompatibleWrapper>()

  useImperativeHandle(
    ref,
    () => ({
      setOffset(offset) {
        if (currentAnimation.current) {
          currentAnimation.current.stop()
        }
        currentAnimation.current = Animated.spring(currentOffset, {
          ...Animated.SpringUtils.makeDefaultConfig(),
          stiffness: 600,
          damping: 120,
          toValue: -offset,
        })

        currentAnimation.current.start(() => {
          currentAnimation.current = null
        })
      },
    }),
    []
  )

  return (
    <Animated.View
      style={{
        flex: 1,
        width: width * 3,
        flexDirection: "row",
        transform: [
          {
            translateX: currentOffset as any,
          },
        ],
      }}
    >
      {children}
    </Animated.View>
  )
})
