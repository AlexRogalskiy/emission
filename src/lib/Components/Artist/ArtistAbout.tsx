import React from "react"
import { createFragmentContainer, graphql } from "react-relay"

import RelatedArtists from "../RelatedArtists"
import Articles from "./Articles"
import Biography from "./Biography"

import { Box, Separator, Spacer } from "@artsy/palette"
import { ArtistAbout_artist } from "__generated__/ArtistAbout_artist.graphql"
import SwitchBoard from "lib/NativeModules/SwitchBoard"
import { get } from "lib/utils/get"
import { CaretButton } from "../Buttons/CaretButton"
import { StickyTabPageScrollView } from "../StickyTabPage/StickyTabPageScrollView"

interface Props {
  artist: ArtistAbout_artist
}

class ArtistAbout extends React.Component<Props> {
  render() {
    return (
      <StickyTabPageScrollView>
        {this.biography()}
        {this.articles()}
        {this.relatedArtists()}
      </StickyTabPageScrollView>
    )
  }

  biography() {
    if (this.props.artist.has_metadata) {
      return (
        <Box>
          <Biography artist={this.props.artist as any} />
          {this.auctionResults()}
          <Separator />
        </Box>
      )
    }
  }

  auctionResults() {
    if (this.props.artist.is_display_auction_link) {
      const url = `/artist/${this.props.artist.slug}/auction-results`
      return (
        <>
          <CaretButton onPress={() => SwitchBoard.presentNavigationViewController(this, url)} text="Auction results" />
          <Spacer mb={3} />
        </>
      )
    }
  }

  articles() {
    if (this.props.artist.articles.edges.length) {
      return (
        <>
          <Box my={3}>
            <Articles articles={this.props.artist.articles.edges.map(({ node }) => node)} />
          </Box>
          <Separator />
        </>
      )
    }
  }

  relatedArtists() {
    const relatedArtistsPresent = get(this.props, p => p.artist.related.artists.edges[0])

    return (
      relatedArtistsPresent && (
        <Box my={3}>
          <RelatedArtists artists={this.props.artist.related.artists.edges.map(({ node }) => node)} />
        </Box>
      )
    )
  }
}

export default createFragmentContainer(ArtistAbout, {
  artist: graphql`
    fragment ArtistAbout_artist on Artist {
      has_metadata: hasMetadata
      is_display_auction_link: isDisplayAuctionLink
      slug
      ...Biography_artist
      related {
        artists: artistsConnection(first: 16) {
          edges {
            node {
              ...RelatedArtists_artists
            }
          }
        }
      }
      articles: articlesConnection(first: 10) {
        edges {
          node {
            ...Articles_articles
          }
        }
      }
    }
  `,
})
