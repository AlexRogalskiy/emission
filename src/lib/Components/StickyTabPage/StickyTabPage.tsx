import { Box, color, Flex, Sans, space, Spacer } from "@artsy/palette"
import { useScreenDimensions } from "lib/utils/useScreenDimensions"
import React, { useMemo, useRef, useState } from "react"
import { TouchableOpacity, View } from "react-native"
import Animated from "react-native-reanimated"
import styled from "styled-components/native"
import { useAnimatedValue } from "./reanimatedHelpers"
import { SnappyHorizontalRail } from "./SnappyHorizontalRail"
import { StickyTabScrollView, TabState } from "./StickyTabScrollView"

interface TabProps {
  initial?: boolean
  title: string
  content: JSX.Element
}

/**
 * This page wrapper encapsulates a disappearing header and sticky tabs each with their own content
 *
 * At the moment all tabs are rendered at all times, as this isn't designed for more than 3 tabs
 * but if we need to have conditional rendering of tab content in the future it should be possible.
 *
 * Each tab optionally consumes a 'scroll view context' which could potentialy contain information
 * about whether the tab is being shown currently etc.
 */
export const StickyTabPage: React.FC<{
  tabs: TabProps[]
  headerContent: JSX.Element
}> = ({ tabs, headerContent }) => {
  const initialTabIndex = useMemo(() => Math.max(tabs.findIndex(tab => tab.initial), 0), [tabs])
  const { width } = useScreenDimensions()
  const activeTabIndex = useAnimatedValue(initialTabIndex)
  const tabStates: TabState[] = useMemo(
    () => {
      return tabs.map((_, index) => ({
        isActive: Animated.eq(index, activeTabIndex),
        shouldLoad: new Animated.Value(index === initialTabIndex ? 1 : 0),
        hasLoaded: new Animated.Value(0),
      }))
    },
    [tabs]
  )
  // trigger staggered loading of tabs
  Animated.useCode(
    () => {
      return Animated.call(
        // collect all shouldLoad and hasLoaded into a single array so we can pluck them out at once
        [...tabStates.map(({ shouldLoad }) => shouldLoad), ...tabStates.map(({ hasLoaded }) => hasLoaded)],
        vals => {
          const shouldsLoads = vals.slice(0, tabStates.length)
          const hasLoadeds = vals.slice(tabStates.length)

          let loadInProgress = false
          for (let tabIndex = 0; tabIndex < tabStates.length; tabIndex++) {
            if (!hasLoadeds[tabIndex] && shouldsLoads[tabIndex]) {
              loadInProgress = true
              break
            }
          }
          if (loadInProgress) {
            return
          }
          // kick the next one off if needed
          for (let tabIndex = 0; tabIndex < tabStates.length; tabIndex++) {
            if (!hasLoadeds[tabIndex] && !shouldsLoads[tabIndex]) {
              tabStates[tabIndex].shouldLoad.setValue(1)
              break
            }
          }
        }
      )
    },
    [tabStates]
  )
  const [headerHeight, setHeaderHeight] = useState<number>(0)

  const headerOffsetY = useAnimatedValue(0)

  const railRef = useRef<SnappyHorizontalRail>()
  const initialRenderTime = useMemo(() => Date.now(), [])

  return (
    <View style={{ flex: 1, position: "relative", overflow: "hidden" }}>
      {/* put tab content first because we want the header to be absolutely positioned _above_ it */}
      <SnappyHorizontalRail ref={railRef} initialOffset={initialTabIndex * width}>
        {tabs.map(({ content }, index) => {
          return (
            <View
              style={{ flex: 1, width }}
              key={index}
              onLayout={() => {
                if (headerHeight > 0) {
                  console.warn("second layout", Date.now() - initialRenderTime)
                }
              }}
            >
              <StickyTabScrollView
                headerHeight={headerHeight}
                content={content}
                headerOffsetY={headerOffsetY}
                tabState={tabStates[index]}
              />
            </View>
          )
        })}
      </SnappyHorizontalRail>
      <Animated.View
        style={{
          width,
          top: 0,
          position: "absolute",
          backgroundColor: color("white100"),
          transform: [{ translateY: headerOffsetY as any }],
        }}
      >
        <View
          onLayout={e => {
            setHeaderHeight(e.nativeEvent.layout.height)
            console.warn("first layout", Date.now() - initialRenderTime)
          }}
        >
          {headerContent}
          <Spacer mb={1} />
        </View>
        <StickyTabBar>
          {tabs.map(({ title }, index) => (
            <StickyTab
              key={title}
              label={title}
              active={tabStates[index].isActive}
              onPress={() => {
                activeTabIndex.setValue(index)
                railRef.current.setOffset(index * width)
                tabStates[index].shouldLoad.setValue(1) // *just* in case. This should already be 1
              }}
            />
          ))}
        </StickyTabBar>
      </Animated.View>
    </View>
  )
}

export const TAB_BAR_HEIGHT = 48

const StickyTabBar = styled(Flex)`
  border-bottom-width: 1;
  border-bottom-color: ${color("black30")};
  height: ${TAB_BAR_HEIGHT};
  flex-direction: row;
  padding-right: ${space(2)};
  padding-left: ${space(2)};
`

const StickyTab: React.FC<{ label: string; active: Animated.Node<number>; onPress(): void }> = ({
  label,
  active,
  onPress,
}) => {
  const [isActive, setIsActive] = useState(false)

  // Animated.useCode(() => Animated.call([active], ([val]) => setIsActive(Boolean(val))), [])

  return (
    <Flex style={{ flex: 1, height: TAB_BAR_HEIGHT }}>
      <TouchableOpacity onPress={onPress}>
        <Box
          style={{
            height: TAB_BAR_HEIGHT,
            alignItems: "center",
            justifyContent: "center",
            borderBottomWidth: 1,
            borderBottomColor: isActive ? color("black100") : color("black30"),
          }}
        >
          <Sans size="3" weight={isActive ? "medium" : "regular"}>
            {label}
          </Sans>
        </Box>
      </TouchableOpacity>
    </Flex>
  )
}
