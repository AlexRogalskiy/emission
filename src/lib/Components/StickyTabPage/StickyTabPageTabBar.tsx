import { Box, color, Flex, Sans, space } from "@artsy/palette"
import React from "react"
import { TouchableOpacity, View } from "react-native"
import { useStickyTabPageContext } from "./StickyTabPage"

const TAB_BAR_HEIGHT = 48

export const StickyTabPageTabBar: React.FC = () => {
  const { tabLabels, activeTabIndex, setActiveTabIndex } = useStickyTabPageContext()
  activeTabIndex.useUpdates()
  return (
    <View
      style={{
        borderBottomWidth: 1,
        borderBottomColor: color("black30"),
        height: TAB_BAR_HEIGHT,
        flexDirection: "row",
        paddingHorizontal: space(2),
      }}
    >
      {tabLabels.map((label, index) => (
        <StickyTab
          key={index}
          label={label}
          active={index === activeTabIndex.current}
          onPress={() => {
            setActiveTabIndex(index)
          }}
        />
      ))}
    </View>
  )
}

export const StickyTab: React.FC<{ label: string; active: boolean; onPress(): void }> = ({
  label,
  active,
  onPress,
}) => {
  return (
    <Flex style={{ flex: 1, height: TAB_BAR_HEIGHT }}>
      <TouchableOpacity onPress={onPress}>
        <Box
          style={{
            height: TAB_BAR_HEIGHT,
            alignItems: "center",
            justifyContent: "center",
            borderBottomWidth: 1,
            borderBottomColor: active ? color("black100") : color("black30"),
          }}
        >
          <Sans size="3" weight={active ? "medium" : "regular"}>
            {label}
          </Sans>
        </Box>
      </TouchableOpacity>
    </Flex>
  )
}
