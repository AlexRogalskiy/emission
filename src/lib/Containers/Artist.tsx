import { Flex, Spacer, Theme } from "@artsy/palette"
import { Artist_artist } from "__generated__/Artist_artist.graphql"
import ArtistAbout from "lib/Components/Artist/ArtistAbout"
import ArtistArtworks from "lib/Components/Artist/ArtistArtworks"
import ArtistHeader from "lib/Components/Artist/ArtistHeader"
import ArtistShows from "lib/Components/Artist/ArtistShows"
import { StickyTabPage } from "lib/Components/StickyTabPage/StickyTabPage"
import { Schema, screenTrack } from "lib/utils/track"
import { ProvideScreenDimensions } from "lib/utils/useScreenDimensions"
import React from "react"
import { ViewProperties } from "react-native"
import { createFragmentContainer, graphql } from "react-relay"

interface Props extends ViewProperties {
  artist: Artist_artist
}

interface State {
  tabs: any[]
}

@screenTrack((props: Props) => ({
  context_screen: Schema.PageNames.ArtistPage,
  context_screen_owner_slug: props.artist.slug,
  context_screen_owner_id: props.artist.internalID,
  context_screen_owner_type: Schema.OwnerEntityTypes.Artist,
}))
export class Artist extends React.Component<Props, State> {
  state = {
    tabs: [],
  }

  componentWillMount = () => {
    this.availableTabs()
  }

  availableTabs = () => {
    const tabs = []
    const artist = this.props.artist
    const displayAboutSection = artist.has_metadata || artist.counts.articles > 0 || artist.counts.related_artists > 0

    if (displayAboutSection) {
      tabs.push({
        title: "About",
        content: <ArtistAbout artist={artist} />,
      })
    }

    if (artist.counts.artworks) {
      tabs.push({
        title: "Artworks",
        initial: true,
        content: <ArtistArtworks artist={artist} />,
      })
    }

    if (artist.counts.partner_shows) {
      tabs.push({
        title: "Shows",
        content: <ArtistShows artist={artist} />,
      })
    }

    this.setState({ tabs })
  }

  render() {
    const { artist } = this.props
    const { tabs } = this.state

    return (
      <Theme>
        <ProvideScreenDimensions>
          <Flex style={{ flex: 1 }}>
            <StickyTabPage
              staticHeaderContent={
                <>
                  <ArtistHeader artist={artist} />
                  <Spacer mb={1} />
                </>
              }
              tabs={tabs}
            />
          </Flex>
        </ProvideScreenDimensions>
      </Theme>
    )
  }
}

export default createFragmentContainer(Artist, {
  artist: graphql`
    fragment Artist_artist on Artist {
      internalID
      slug
      has_metadata: hasMetadata
      counts {
        artworks
        partner_shows: partnerShows
        related_artists: relatedArtists
        articles
      }
      ...ArtistHeader_artist
      ...ArtistAbout_artist
      ...ArtistShows_artist
      ...ArtistArtworks_artist
    }
  `,
})
