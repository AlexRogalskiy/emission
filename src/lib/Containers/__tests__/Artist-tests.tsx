import { Artist_artist } from "__generated__/Artist_artist.graphql"
import { StickyTab } from "lib/Components/StickyTabPage/StickyTabPageTabBar"
import Artist from "lib/Containers/Artist"
import { renderRelayTree } from "lib/tests/renderRelayTree"
import React from "react"
import "react-native"
import { graphql } from "react-relay"

jest.unmock("react-relay")

describe("availableTabs", () => {
  const getWrapper = async (artist: Omit<Artist_artist, " $fragmentRefs">) =>
    await renderRelayTree({
      Component: (props: any) => {
        return <Artist isPad={false} artist={{ ...artist }} {...props} />
      },
      query: graphql`
        query ArtistTestsQuery($isPad: Boolean!) @raw_response_type {
          artist(id: "andy-warhol") {
            ...Artist_artist
          }
        }
      `,
      variables: {
        isPad: false,
      },
      mockData: {
        artist,
      },
    })

  it("returns nothing if artist has no metadata, shows, or works", async () => {
    const wrapper = await getWrapper(ArtistFixture as any)
    expect(wrapper.find(StickyTab).length).toEqual(0)
  })

  it("returns About tab if artist has metadata", async () => {
    const wrapper = await getWrapper({ ...ArtistFixture, has_metadata: true } as any)
    expect(wrapper.find(StickyTab).length).toEqual(1)
    expect(wrapper.find(StickyTab).text()).toContain("About")
  })

  it("returns About tab if artist has articles", async () => {
    const wrapper = await getWrapper({
      ...ArtistFixture,
      counts: {
        artworks: 0,
        partner_shows: 0,
        related_artists: 0,
        articles: 1,
        follows: 0,
      },
      articles: {
        edges: [
          {
            node: {
              thumbnail_title: "Record-Breaking Ed Ruscha Brings in $52.4 Million at Christie’s",
              href: "/article/artsy-editorial-record-breaking-ruscha-brings-524-million-christies",
              author: {
                name: "Artsy Editorial",
              },
              thumbnail_image: {
                url:
                  "https://artsy-media-uploads.s3.amazonaws.com/7AcSnA5B-WhzwL_lvogGTA%2Fcustom-Custom_Size___Lot+6%2C+Ruscha%2C+Hurting+the+Word+Radio+%232.jpg",
              },
            },
          },
        ],
      },
    } as any)
    expect(wrapper.find(StickyTab).length).toEqual(1)
    expect(wrapper.find(StickyTab).text()).toContain("About")
  })

  it("returns Shows tab if artist has shows", async () => {
    const wrapper = await getWrapper({
      ...ArtistFixture,
      counts: {
        artworks: 0,
        partner_shows: 2,
        related_artists: 0,
        articles: 0,
        follows: 0,
      },
      currentShows: {
        edges: [
          {
            node: {
              id: "U2hvdzo1OGE1M2YzYzc2MjJkZDQxNmY3YTNjNGQ=",
              slug: "leeum-samsung-museum-of-art-leeum-collection-beyond-space",
              href: "/show/leeum-samsung-museum-of-art-leeum-collection-beyond-space",
              is_fair_booth: false,
              cover_image: {
                url: "https://d32dm0rphc51dk.cloudfront.net/vlXf0LFyHT9wTe7ODTMuJg/large.jpg",
              },
              kind: "group",
              name: "Leeum Collection: Beyond Space",
              exhibition_period: "Aug 19, 2014 – Dec 31, 2999",
              status_update: null,
              status: "running",
              partner: {
                __typename: "Partner",
                name: "Leeum, Samsung Museum of Art",
              },
              location: {
                city: "Seoul ",
              },
            },
          },
        ],
      },
    } as any)
    expect(wrapper.find(StickyTab).length).toEqual(1)
    expect(wrapper.find(StickyTab).text()).toContain("Shows")
  })

  it("returns all three tabs if artist has metadata, works, and shows", async () => {
    const wrapper = await getWrapper({
      ...ArtistFixture,
      counts: {
        artworks: 2,
        partner_shows: 2,
        related_artists: 0,
        articles: 0,
        follows: 0,
      },
      has_metadata: true,
      artworks: {
        pageInfo: {
          hasNextPage: true,
          startCursor: "YXJyYXljb25uZWN0aW9uOjA=",
          endCursor: "YXJyYXljb25uZWN0aW9uOjA=",
        },
        edges: [
          {
            cursor: "YXJyYXljb25uZWN0aW9uOjA=",
            node: {
              slug: "andy-warhol-chanel-no-5-set-of-4",
              id: "QXJ0d29yazo1OTdmZjliYTJhODkzYTI1ZDgyNDRiYWM=",
              image: {
                aspectRatio: 0.77,
                url: "https://d32dm0rphc51dk.cloudfront.net/b3nN0Y0pvZXGE7DnktZttA/large.jpg",
                aspect_ratio: 0.77,
              },
              title: "Chanel No. 5 (set of 4)",
              date: "1997",
              sale_message: "Contact For Price",
              is_biddable: false,
              is_acquireable: false,
              is_offerable: false,
              sale: null,
              sale_artwork: null,
              artists: [
                {
                  name: "Andy Warhol",
                },
              ],
              partner: {
                __typename: "Partner",
                name: "EHC Fine Art",
              },
              href: "/artwork/andy-warhol-chanel-no-5-set-of-4",
            },
          },
        ],
      },
    } as any)
    expect(wrapper.find(StickyTab).length).toEqual(3)
    expect(
      wrapper
        .find(StickyTab)
        .at(0)
        .text()
    ).toContain("About")
    expect(
      wrapper
        .find(StickyTab)
        .at(1)
        .text()
    ).toContain("Artworks")
    expect(
      wrapper
        .find(StickyTab)
        .at(2)
        .text()
    ).toContain("Shows")
  })
})

const ArtistFixture = {
  internalID: "4d8b92b34eb68a1b2c0003f4",
  slug: "andy-warhol",
  hasMetadata: true,
  counts: {
    artworks: 0,
    partner_shows: 0,
    related_artists: 0,
    articles: 0,
    follows: 0,
  },
  id: "QXJ0aXN0OjRkOGI5MmIzNGViNjhhMWIyYzAwMDNmNA==",
  isFollowed: false,
  name: "Andy Warhol",
  nationality: "American",
  birthday: "1928",
  has_metadata: false,
  is_display_auction_link: true,
  bio: "American, 1928-1987, Pittsburgh, PA, United States, based in New York, NY, United States",
  blurb:
    "Obsessed with [celebrity](/collection/andy-warhol-celebrity-portraits), consumer culture, and mechanical (re)production, [Pop artist](/gene/pop-art) Andy Warhol created some of the most iconic images of the 20th century. As famous for his quips as for his art—he variously mused that “art is what you can get away with” and “everyone will be famous for 15 minutes”—Warhol drew widely from popular culture and everyday subject matter, creating works like his [_32 Campbell's Soup Cans_](/collection/andy-warhol-campbells-soup-can) (1962), [Brillo pad box sculptures](/collection/andy-warhol-brillo-boxes), and portraits of [Marilyn Monroe](/collection/andy-warhol-marilyn-monroe), using the medium of silk-screen printmaking to achieve his characteristic hard edges and flat areas of color. Known for his cultivation of celebrity, Factory studio (a radical social and creative melting pot), and avant-garde films like _Chelsea Girls_ (1966), Warhol was also a mentor to artists like [Keith Haring](/artist/keith-haring) and [Jean-Michel Basquiat](/artist/jean-michel-basquiat). His Pop sensibility is now standard practice, taken up by major contemporary artists [Richard Prince](/artist/richard-prince), [Takashi Murakami](/artist/takashi-murakami), and [Jeff Koons](/artist/jeff-koons), among countless others.",
  related: {
    artists: {
      edges: [],
    },
  },
  artworks: {
    edges: [],
    pageInfo: {
      hasNextPage: false,
      startCursor: "",
      endCursor: "",
    },
  },
  articles: {
    edges: [],
  },
  currentShows: {
    edges: [],
  },
  upcomingShows: {
    edges: [],
  },
  pastSmallShows: {
    edges: [],
  },
  pastLargeShows: {
    edges: [],
  },
}
