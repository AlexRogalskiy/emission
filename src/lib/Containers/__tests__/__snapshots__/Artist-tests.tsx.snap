// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layout works as expected with no tabs 1`] = `
<RCTScrollView
  automaticallyAdjustContentInsets={false}
  scrollsToTop={true}
>
  <View>
    <View
      style={
        Object {
          "paddingLeft": 40,
          "paddingRight": 40,
        }
      }
    >
      <Header
        artist={
          Object {
            "birthday": null as any,
            "counts": Object {
              "articles": 0,
              "artworks": 0,
              "partner_shows": 0,
            },
            "current_shows": null as any,
            "has_metadata": false,
            "id": null as any,
            "internalID": null as any,
            "name": null as any,
            "nationality": null as any,
            "past_large_shows": null as any,
            "past_small_shows": null as any,
            "upcoming_shows": null as any,
          }
        }
      />
      <View
        style={
          Object {
            "paddingTop": 30,
          }
        }
      />
    </View>
  </View>
</RCTScrollView>
`;

exports[`layout works as expected with one tab 1`] = `
<RCTScrollView
  automaticallyAdjustContentInsets={false}
  scrollsToTop={true}
>
  <View>
    <View
      style={
        Object {
          "paddingLeft": 40,
          "paddingRight": 40,
        }
      }
    >
      <Header
        artist={
          Object {
            "birthday": null as any,
            "counts": Object {
              "articles": 0,
              "artworks": 0,
              "partner_shows": 0,
            },
            "current_shows": null as any,
            "has_metadata": true,
            "id": null as any,
            "internalID": null as any,
            "name": null as any,
            "nationality": null as any,
            "past_large_shows": null as any,
            "past_small_shows": null as any,
            "upcoming_shows": null as any,
          }
        }
      />
      <View
        style={
          Object {
            "paddingTop": 30,
          }
        }
      >
        <About
          artist={
            Object {
              "birthday": null as any,
              "counts": Object {
                "articles": 0,
                "artworks": 0,
                "partner_shows": 0,
              },
              "current_shows": null as any,
              "has_metadata": true,
              "id": null as any,
              "internalID": null as any,
              "name": null as any,
              "nationality": null as any,
              "past_large_shows": null as any,
              "past_small_shows": null as any,
              "upcoming_shows": null as any,
            }
          }
        />
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`layout works as expected with three tabs 1`] = `
<RCTScrollView
  automaticallyAdjustContentInsets={false}
  scrollsToTop={true}
>
  <View>
    <View
      style={
        Object {
          "paddingLeft": 40,
          "paddingRight": 40,
        }
      }
    >
      <Header
        artist={
          Object {
            "birthday": null as any,
            "counts": Object {
              "artworks": 2,
              "partner_shows": 1,
            },
            "current_shows": null as any,
            "has_metadata": true,
            "id": null as any,
            "internalID": null as any,
            "name": null as any,
            "nationality": null as any,
            "past_large_shows": null as any,
            "past_small_shows": null as any,
            "upcoming_shows": null as any,
          }
        }
      />
      <View>
        <SwitchView
          onSelectionChange={[Function]}
          selectedIndex={1}
          style={
            Object {
              "marginBottom": 30,
              "marginLeft": 0,
              "marginRight": 0,
              "marginTop": 30,
            }
          }
          titles={
            Array [
              "ABOUT",
              "WORKS",
              "SHOWS",
            ]
          }
        />
        <View>
          <Artworks
            artist={
              Object {
                "birthday": null as any,
                "counts": Object {
                  "artworks": 2,
                  "partner_shows": 1,
                },
                "current_shows": null as any,
                "has_metadata": true,
                "id": null as any,
                "internalID": null as any,
                "name": null as any,
                "nationality": null as any,
                "past_large_shows": null as any,
                "past_small_shows": null as any,
                "upcoming_shows": null as any,
              }
            }
          />
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
